### PR TITLE
[ENH] Add n_elements_ attribute to masker classes

### LINKED ISSUE #3296: Add `n_elements_` attribute to masker objects
It would be nice to have an attribute in masker objects that indicates the number of regions/voxels in the mask.

I am mostly opening this to gauge interest. In the past, I have lamented the fact that I don't have ready access to the number of regions/voxels in different types of maskers, and I thought it might be worth adding as an attribute.

### Benefits to the change

I think this would primarily benefit developers who rely on Nilearn, rather than users. The main benefit would be that it would be easier to preallocate arrays for transformed images without having to directly check the mask images.

### Pseudocode for the new behavior, if applicable

```python
from nilearn import datasets, maskers
dataset = datasets.fetch_development_fmri(n_subjects=1)
func_filename = dataset.func[0]

nifti_masker = maskers.NiftiMasker(mask_strategy='epi')
nifti_masker.fit(func_filename)
mask_img = nifti_masker.mask_img_
print(nifti_masker.n_elements_)
# 24256
```

### PR DESCRIPTION
**Pull request by tsalo**
Closes #3296.

Changes proposed in this pull request:

- Add an `n_elements_` attribute to each of the masker classes.
- Document any post-fit attributes in the masker class docstrings.
- Expand masker tests to check the post-fit attributes.
- Some minor improvements to coding style in the masker code files.
- Replace relative imports in multi_nifti_masker.py and nifti_masker.py with absolute ones.
- Check that mask is binary/valid in fit, rather than just in transform, for NiftiMasker and MultiNiftiMasker.
    - Still allow empty mask at this stage, though it will fail at transform.
- Apply consistent, black-like formatting to the masker files (except using single quotes, of course).

### CODE REVIEW

File: nilearn/maskers/multi_nifti_masker.py
Comment (bthirion) [inline nilearn/maskers/multi_nifti_masker.py:279]: Silly question maybe, but is the mask always binary ? We need to check whether the mask is forced to be binary.

File: nilearn/maskers/tests/test_nifti_maps_masker.py
Comment (bthirion) [inline nilearn/maskers/tests/test_nifti_maps_masker.py:4]: typo

Comment (tsalo) [reply nilearn/maskers/tests/test_nifti_maps_masker.py:4]: Thanks for catching that!

File: nilearn/maskers/multi_nifti_masker.py
Comment (tsalo) [reply nilearn/maskers/multi_nifti_masker.py:279]: I assumed it was, but it looks like that is only checked at `transform` via `_load_mask_img`. Should I add a check in `fit` as well?

Comment (tsalo) [reply nilearn/maskers/multi_nifti_masker.py:273]: I swapped out `get_data` with `_load_mask_img` in `fit`, which will check that the mask is binary.

Comment (tsalo) [reply nilearn/maskers/multi_nifti_masker.py:273]: It looks like the tests `test_nifti_masker.py::test_compute_brain_mask` and `test_multi_nifti_masker.py::test_compute_multi_gray_matter_mask` have multiple checks that explicitly allow empty masks. @NicolasGensollen it looks like you added these tests in #2904. Were they just smoke tests, or do we want to allow empty masks at `fit`?

Comment (bthirion) [reply nilearn/maskers/multi_nifti_masker.py:302]: My gut feeling is that we show allow empty masks.

Comment (tsalo) [reply nilearn/maskers/multi_nifti_masker.py:302]: I think there's a typo in your comment. Are you saying "should allow" or "shouldn't allow"?

Comment (bthirion) [reply nilearn/maskers/multi_nifti_masker.py:302]: Sorry: we *should* allow.

Comment (tsalo) [reply nilearn/maskers/multi_nifti_masker.py:302]: Ah, okay. What would the expected behavior be in that case? Should (1) fit pass and transform fail, (2) fit and transform pass, but transform returns an array of NaNs, or (3) fit and transform pass, but transform returns an empty array?

Comment (bthirion) [reply nilearn/maskers/multi_nifti_masker.py:302]: The last one makes more sense to me.

Comment (tsalo) [reply nilearn/maskers/multi_nifti_masker.py:302]: I think most of that will need to be tackled in a separate PR, but for this one I can ensure that an exception is not raised when the mask is empty.

File: nilearn/maskers/nifti_spheres_masker.py
Comment (bthirion) [inline nilearn/maskers/nifti_spheres_masker.py:157]:
```suggestion
    seeds : :obj:`list` of triplets of coordinates in native space
```

**Review by bthirion — COMMENTED** [review]
LGTM. This is mostly a documentation cleaning PR.

File: nilearn/maskers/multi_nifti_masker.py
Comment (tsalo) [reply nilearn/maskers/multi_nifti_masker.py:51]: I was just consolidating imports, and there were several from `_utils`, so it seemed more efficient to just use `_utils.fill_doc` instead of import `fill_doc` separately.

**Review by bthirion — APPROVED** [review]
LGTM, thx.

File: nilearn/maskers/nifti_masker.py
Comment (ymzayek) [inline nilearn/maskers/nifti_masker.py:417]: 'list' in line 414 can be linked with :obj: role

File: nilearn/maskers/nifti_masker.py
Comment (ymzayek) [inline nilearn/maskers/nifti_masker.py:509]: idem regarding `array-like`.

File: nilearn/maskers/nifti_spheres_masker.py
Comment (ymzayek) [inline nilearn/maskers/nifti_spheres_masker.py:411]: idem regarding `array-like`.

File: nilearn/maskers/nifti_spheres_masker.py
Comment (ymzayek) [inline nilearn/maskers/nifti_spheres_masker.py:450]: idem regarding `array-like`.

File: nilearn/maskers/nifti_masker.py
Comment (ymzayek) [inline nilearn/maskers/nifti_masker.py:410]:
```suggestion
        """Compute the mask corresponding to the data.
```

File: nilearn/maskers/nifti_masker.py
Comment (ymzayek) [inline nilearn/maskers/nifti_masker.py:500]:
```suggestion
        """Apply mask, spatial and temporal preprocessing.
```

File: nilearn/maskers/nifti_labels_masker.py
Comment (ymzayek) [inline nilearn/maskers/nifti_labels_masker.py:465]: `imgs` is defined as 3D or 4D but in the description it says "It must boil down to a 4D image with scans number as last dimension."

File: nilearn/maskers/nifti_labels_masker.py
Comment (ymzayek) [inline nilearn/maskers/nifti_labels_masker.py:505]: idem regarding discrepancy of definitions and description of imgs

File: nilearn/maskers/nifti_masker.py
Comment (ymzayek) [inline nilearn/maskers/nifti_masker.py:506]: idem regarding discrepancy of definitions and description of imgs

File: nilearn/maskers/nifti_spheres_masker.py
Comment (ymzayek) [inline nilearn/maskers/nifti_spheres_masker.py:409]: idem regarding discrepancy of definitions and description of imgs

**Review by ymzayek — COMMENTED** [review]
Overall LGTM @tsalo ! Just minor things for consistency.

File: nilearn/maskers/nifti_labels_masker.py
Comment (tsalo) [reply nilearn/maskers/nifti_labels_masker.py:467]: I think the one in `multi_nifti_masker` would be a list of objects, with the entries being equivalent to the parameter here.

Comment (tsalo) [reply nilearn/maskers/nifti_labels_masker.py:467]: I don't know about making `array-like` more specific. I think `array-like` means that it could be a numpy array, a list of lists, or possibly some other objects with array-like organization. I don't know what specific checks/conversions are done, though, so I am not sure about specifically using `numpy

Comment (ymzayek) [reply nilearn/maskers/nifti_labels_masker.py:467]: Ok fine to leave as is then!

File: nilearn/maskers/nifti_labels_masker.py
Comment (tsalo) [reply nilearn/maskers/nifti_labels_masker.py:465]: Good point. I checked `NiftiMasker`, `NiftiLabelsMasker`, and `NiftiMapsMasker`, and all three can handle 3D niimgs just fine. They all just add a singleton dimension to produce a 2D output. I'm updating the docstrings to reflect that.

File: nilearn/maskers/nifti_labels_masker.py
Comment (tsalo) [reply nilearn/maskers/nifti_labels_masker.py:510]: I think it may be, but the problem is that `confounds` is used a lot outside of `maskers`, so it might be good to address in a future PR.

**Review by ymzayek — APPROVED** [review]
Thanks @tsalo

**Comment by tsalo** [comment]
Thanks @ymzayek! Would you prefer that I merge this soon, and you could add in the new attribute to the new classes in #3237 yourself, or that I wait until #3237 is merged and then add the attribute to those classes here?

**Comment by ymzayek** [comment]
@tsalo I would say better to go ahead and merge this one and then I can handle the new attribute in #3237. I have to revisit it anyways and I think it might take some time until it is ready.

**Comment by tsalo** [comment]
Okay, I'll merge now then. Thanks!

**Pull request merged**
